pkg/drivers/krunkit/krunkit.go

```diff
@@ -0,0 +1,562 @@
+	}
+
+	log.Debugf("executing: krunkit %s", strings.Join(args, " "))
+	cmd := exec.Command(driverName, args...)
+
+	// Create krunkit in a new process group, so minikube caller can use killpg
+	// to terminate the entire process group without harming the krunkit process.
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	logfile, err := d.openLogfile()
+	if err != nil {
+		return fmt.Errorf("failed to open krunkit logfile: %w", err)
+	}
+	defer logfile.Close()
+	cmd.Stderr = logfile
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	return process.WritePidfile(d.pidfilePath(), cmd.Process.Pid)
+}
+
+func (d *Driver) openLogfile() (*os.File, error) {
+	logfile := d.ResolveStorePath(logFileName)
+	return os.OpenFile(logfile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+}
+
+func (d *Driver) setupIP(mac string) error {
+	var err error
+	getIP := func() error {
+		d.IPAddress, err = pkgdrivers.GetIPAddressByMACAddress(mac)
+		if err != nil {
+			return errors.Wrap(err, "failed to get IP address")
+		}
+		return nil
+	}
+	// Implement a retry loop because IP address isn't added to dhcp leases file immediately
+	for i := 0; i < 60; i++ {
+		log.Debugf("Attempt %d", i)
+		err = getIP()
+		if err == nil {
+			break
+		}
+		time.Sleep(2 * time.Second)
+	}
+
+	if err == nil {
+		log.Debugf("IP: %s", d.IPAddress)
+		return nil
+	}
+	if !isBootpdError(err) {
+		return errors.Wrap(err, "IP address never found in dhcp leases file")
+	}
+	if unblockErr := firewall.UnblockBootpd(); unblockErr != nil {
+		klog.Errorf("failed unblocking bootpd from firewall: %v", unblockErr)
+		exit.Error(reason.IfBootpdFirewall, "ip not found", err)
+	}
+	out.Styled(style.Restarting, "Successfully unblocked bootpd process from firewall, retrying")
+	return fmt.Errorf("ip not found: %v", err)
+}
+
+func isBootpdError(err error) bool {
+	return strings.Contains(err.Error(), "could not find an IP address")
+}
+
+func (d *Driver) Stop() error {
+	if err := d.stopKrunkit(); err != nil {
+		return err
+	}
+	return d.VmnetHelper.Stop()
+}
+
+func (d *Driver) Kill() error {
+	if err := d.killKrunkit(); err != nil {
+		return err
+	}
+	return d.VmnetHelper.Kill()
+}
+
+func (d *Driver) Remove() error {
+	s, err := d.GetState()
+	if err != nil {
+		return errors.Wrap(err, "get state")
+	}
+	if s == state.Running {
+		if err := d.Kill(); err != nil {
+			return errors.Wrap(err, "kill")
+		}
+	}
+	return nil
+}
+
+func (d *Driver) Restart() error {
+	s, err := d.GetState()
+	if err != nil {
+		return err
+	}
+
+	if s == state.Running {
+		if err := d.Stop(); err != nil {
+			return err
+		}
+	}
+	return d.Start()
+}
+
+func (d *Driver) StartDocker() error {
+	return fmt.Errorf("hosts without a driver cannot start docker")
+}
+
+func (d *Driver) StopDocker() error {
+	return fmt.Errorf("hosts without a driver cannot stop docker")
+}
+
+func (d *Driver) GetDockerConfigDir() string {
+	return ""
+}
+
+func (d *Driver) Upgrade() error {
```

Comment:
I am not sure why there is this func at all, we dont seem to have it in other drivers, except qemu and vfkit... could be just un-used code?

Reply:
Removed since it is unused and not part of the drivers.Driver interface.